### DISCUSSION TITLE
Visiting blocking handler from Browser blocks whole route

### DISCUSSION THREAD

**Discussion by Geo-W** [tokio-rs/axum#2188] [Q&A]
- [*] I have looked for existing issues (including closed) about this
- [*] I already read https://ryhl.io/blog/async-what-is-blocking/ and got that blocking code shouldn't appear in async function, but still feel confusing about the phenomena.

### Version
0.6.20

### Platform

64-bit Windows 11 Home 21H2

### Crates
```toml
axum = { version = "0.6.20", features = ["macros"]}
tokio = { version = "1.32.0", features = ["full"]}
```

also tested on
```
hyper = { version = "1.0.0-rc.4", features = ["full"] }
tokio = { version = "1", features = ["full"] }
http-body-util = "0.1.0-rc.3"
hyper-util = { git = "https://github.com/hyperium/hyper-util.git" }
``` 

### Description

I tried this code:

```rust 
use tokio;
use axum::{response::Html, routing::get, Router};
use std::net::SocketAddr;


#[tokio::main(flavor = "multi_thread", worker_threads = 16)]
async fn main() {

    let app = Router::new()
        .route("/blocking/", get(get_block))
        .route("/nonblock/", get(non_block));

    let addr = SocketAddr::from(([0, 0, 0, 0], 2666));
    axum::Server::bind(&addr)
        .serve(app.into_make_service())
        .await
        .unwrap();
}

async fn get_block() -> Html<&'static str> {
    dbg!(std::thread::current().id());
    let mut i = 1;
	for _ in (0..=599999999){
		i+=1;
	}
    Html("<h1>blocking!</h1>")
}

async fn non_block() -> Html<&'static str> {
    dbg!(std::thread::current().id());
    Html("<h1>Non block!</h1>")
}

```

I expected to see this happen: 
> Visiting blocking handler and non-block handler at the same time, the blocking page wait for the calculation to finish, and the non-block handler return immediately. 

Instead, this happened: 
> When **visiting `http://localhost:2666/blocking/` from the browser**, it blocks all of the following requests no matter they are from the browser or command line `curl`, waiting for blocking to finish then starts handling the following. This problem doesn't exist in actix-web.
> But it works well when visiting `http://localhost:2666/blocking/` with `curl`, and all the following requests are handled properly without waiting for the blocking thread to finish. 
> Test with python `requests` & `threading` works, start 1 blocking request and 15 non-block request, output all non-block first then wait for blocking to finish. 
> Test with `rewrk -c 256 -t 12 -d 12 -h http://127.0.0.1:2666/blocking/` also works pretty well.

**Comment by davidpdrsn** [comment]
Welcome to async rust 😅 that’s just how tokio works and there is nothing axum can do about that. Basically you never wanna block.

See https://ryhl.io/blog/async-what-is-blocking/#:~:text=When%20writing%20async%20Rust%2C%20the,from%20swapping%20the%20current%20task”.

**Reply by Geo-W** [reply]
Yep i got that.  
But what makes me pretty curious is this issue varies for different request agents. It is pretty strange.
